BUG: array2string does not add signs for positive integers. Fixes #24181

paulreece opened the pull request:
This was made to fix [#24181](https://github.com/numpy/numpy/issues/24181).

In this the user described wanting to see similar  behavior in `array2string` for integers as it does for floats when a `sign` argument is given.  I implemented a fix that adds this functionality to the method.

```Python
import numpy as np
arr = np.array([-2, 0, 3])
print(np.array2string(arr, sign='+'))
print(np.array2string(arr.astype(float), sign='+'))

BEFORE FIX:
# output:
# [-2  0  3]
# [-2. +0. +3.]

AFTER FIX:
# output:
# [-2  +0  +3]
# [-2. +0. +3.]
```

paulreece committed:
BUG: array2string does not add signs for positive integers. Fixes #24181.

rkern commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -524,6 +524,14 @@ def test_refcount(self):
+        a = np.array([-2, 0, 3])
+        assert_equal(
+            np.array2string(a, sign='+'),
+            '[+-2 +0 +3]'
```
The semantics are that `+` is only added if the value is positive.

```suggestion
            '[-2 +0 +3]'
```

The results with `sign='-'` and `sign=' '` also need to be tested.

paulreece replied:
@rkern Apologies for the oversight, thank you for the feedback.  I just submitted a fix that implements the desired behavior.

paulreece replied:
Adding tests for sign='-' and sign=' '

rkern commented on numpy/core/arrayprint.py:
```diff
@@ -1220,15 +1220,20 @@ def format_float_positional(x, precision=None, unique=True,
 class IntegerFormat:
-    def __init__(self, data):
+    def __init__(self, data, sign=False):
+        if isinstance(sign, bool):
```
Why do we accept a `bool` here? Because `sign` goes to both `IntegerFormat` and `FloatingFormat`, the accepted values need to be the same. They are exactly just those objects in the set `{None, '+', ' ', '-'}`.

rkern replied:
Ah, I see you got it from `FloatingFormat`, which accepts it due to backwards compatibility concerns. I don't think that will be an issue for `IntegerFormat`.

paulreece replied:
Got rid of the `bool` check.

rkern reviewed (commented):
Thank you for taking a look at this, but I'm afraid this doesn't fix the reported issue. See the individual comments below.

rkern commented on numpy/core/arrayprint.py:
```diff
@@ -1220,15 +1220,21 @@ def format_float_positional(x, precision=None, unique=True,
 
     def __call__(self, x):
+        if self.sign == '+':
+            return '+' + str(x) if x >= 0 else self.format % x
```
To convert `x` to a string, we do need to do the `self.format % x` in both cases, rather than using `str()`.

paulreece replied:
Added in a conditional in `self.format`, if it receives '+' or '  ' as the `sign` it will format it as `'%{}s'.format(max_str_len)` instead.

rkern commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -524,6 +524,22 @@ def test_refcount(self):
+        assert_equal(
+            np.array2string(a, sign=' '),
+            '[-2  0  3]'
+        )
```
To tell the intended difference between `sign='-'` and `sign=' '`, use an all-positive array. See the behavior for floating point inputs as a reference.

```python
>>> np.array2string(np.array([2., 0., 3.]), sign='-')
'[2. 0. 3.]'

>>> np.array2string(np.array([2., 0., 3.]), sign=' ')
'[ 2.  0.  3.]'
```

paulreece replied:
Awesome, I will fix all of the suggested changes.  Grateful for the feedback!

paulreece replied:
Fixed the tests in order to demonstrate the desired behavior.

paulreece commented:
@seberg I just pushed up a commit implementing your suggestion of using modern F-string syntax.  I also the branching.  Thanks for your feedback!

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1220,16 +1221,16 @@ def format_float_positional(x, precision=None, unique=True,
         if data.size > 0:
-            max_str_len = max(len(str(np.max(data))),
+            self.max_str_len = max(len(str(np.max(data))),
                               len(str(np.min(data))))
```
This is incorrect now when the sign is forced and the value is larger than 0 you need an additional character!  (Would be nice to add a test.)

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1220,16 +1221,16 @@ def format_float_positional(x, precision=None, unique=True,
 
     def __call__(self, x):
-        return self.format % x
+        return f"{x:{self.sign}{self.max_str_len}d}"
```
I think you should keep the old logic and define the `self.format`, then only change it to `self.format.format(x)`.  Maybe `self.format.format(int(x))` is actually better/faster (only if you can test that it is).

seberg reviewed (commented):
The `sign` option explicitly mentions "floating-point" types, which is incorrect now and should be adapted.  Could add a `.. versionchanged:: 2.0` directive there also.

paulreece commented:
My comments on the review say pending so I'm adding them here too just in case they aren't rendered

- Implemented all 6 tests for mixed negative/positive array and Non-Negative array
- Implemented @seberg's design suggestions.
- Locally I did add some conditional logic that would add an extra character if a sign was present and the value was >0 but this resulted in adding extra spaces '[ -2  +0  +3]' . I'm wondering if this extra character is needed since this was already rendering signs for negative values before no problem. Just leaving the max_str_len as is results in '[-2 +0 +3]'. Isn't this the desired format, or have I misunderstood? I have pushed up my changes for review

seberg commented:
> comments on the review say pending

You probably clicked on "start a review" that bunches many comments with an additional review one.  They don't show up (there is a green button to finish on the "files" tab).  Or just delete the comments and post them again.

> Isn't this the desired format, or have I misunderstood? I have pushed up my changes for review

Yes, you are right that you don't want those to change.  What happns to `np.array([30, -1, 40])` in the three cases (only with the sign, the positive number printing requires more characters than the negative one).

paulreece commented:
Here's the REPL output from my locally built version in each case for the array:
```Python
>>> import numpy as np
>>> np.array2string(np.array([30, -1, 40]), sign='+')
'[+30 -1 +40]'
>>> np.array2string(np.array([30, -1, 40]), sign='-')
'[30 -1 40]'
>>> np.array2string(np.array([30, -1, 40]), sign=' ')
'[ 30 -1  40]'
```

seberg commented:
Yeah, OK, its hard to see why the above isn't what we want unless we make it a 2-D array like:
`np.array([[10, -1, 1], [1, 1, 10]])`.  The space taken by `-1` in your first and last example is 2, while it is 3 for the the positive numbers.  It should be 3 also for the -1.

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1218,19 +1228,19 @@ def format_float_positional(x, precision=None, unique=True,
         else:
             max_str_len = 0
-        self.format = '%{}d'.format(max_str_len)
+        self.format = lambda x: f"{x:{sign}{max_str_len}d}"
```
I still think that `self.format = f"{{:{sign}{max_str_len}d}}"` makes more sense (also dropping the lambda).  Why interpolate `sign` and `max_str_len` potentially thousands of times?  (yes, a bit of a micro-optimization, do you prefer the lambda?).  (EDIT: no need for `x`, we will format positionally)

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1218,19 +1228,19 @@ def format_float_positional(x, precision=None, unique=True,
             max_str_len = max(len(str(np.max(data))),
                               len(str(np.min(data))))
+            if sign in ['+', ' ']:
+                max_str_len += 1
```
Given an array with all negative numbers, this will cause `sign="-"` to be different from `sign="+"`, which I don't think is wanted.  The additional space is needed for _positive_ values only.

paulreece commented:
Awesome feedback as usual!  Just implemented both comments outlined above.  
- I have the f-string memoizing now as suggested to reduce amount of interpolation.  
- I also added in a check in the `__call__` dunder method to check if the integer is negative and if the `sign` is `'+'` or 
`' '`.  If both evaluate to true it removes a space from the formatted integer.   I figured this would be more performant with the memoizing I just added than adjusting the `max_str_len` variable itself.  However if that is preferred I'm happy to implement that instead.  
- Lastly, I implemented tests for all negative 1d and 2d arrays.

seberg commented:
You have two things:
1. Figure out how much space is necessary for the number/string which takes the *most* space.  This is used for alignment.
2. Format the final integer.

The problem I pointed to is in the first step, but you are trying to fix this up in the second now, that doesn't make sense.  The string returned should be exactly `max_str_length` and you have to make sure that is correct for any input by fixing the definition of `max_str_length`.

paulreece commented:
Okay I think I implemented it as you wanted this time.  I A/B'd all scenarios with the Float version to make sure behavior was the same in both.  Note: I have it reassigning the `' '` to `'-'` if all values aren't positive because it then gives you spacing behavior like it does when you print a Float array with the same values.   I put an example below.
```Python
>>> import numpy as np
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign=' '))
[[10. -1.  1.  1.]
 [10. 10. 10. 10.]]
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign='-'))
[[10. -1.  1.  1.]
 [10. 10. 10. 10.]]
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign='+'))
[[+10.  -1.  +1.  +1.]
 [+10. +10. +10. +10.]]
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign=' '))
[[10 -1  1  1]
 [10 10 10 10]]
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign='-'))
[[10 -1  1  1]
 [10 10 10 10]]
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign='+'))
[[+10  -1  +1  +1]
 [+10 +10 +10 +10]]
```

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1218,19 +1228,27 @@ def format_float_positional(x, precision=None, unique=True,
         if data.size > 0:
-            max_str_len = max(len(str(np.max(data))),
-                              len(str(np.min(data))))
+            data_max = len(str(np.max(data)))
+            data_min = len(str(np.min(data)))
+            max_str_len = max(data_max,
+                              data_min)
+            if (sign == '+' 
+                and (np.all(data >= 0)
+                or (data.ndim > 1 and (data_max == data_min)))):
+                max_str_len += 1
+            elif sign == ' ' and np.all(data >= 0):
+                max_str_len += 1
+            elif sign == ' ':
+                sign = '-'
```
```suggestion
            data_max = np.max(data)
            data_max_str_len = len(str(data_max))
            if data_max >= 0 and sign in "+ ":
                data_max_str_len += 1

            max_str_len = max(data_max_str_len,
                              str(len(np.min(data))))   
```
When the maximum value is 0 or positive it needs one additional character when printing, that is supposed to be accounted for.  This doesn't matter if there are negative numbers which take more space, but it does matter if there are some negative numbers!

Maybe you can think of a test that covers where this diverges from what you have?

We don't actually support `sign=None`, so might as well make it `sign="-"` in the signature.

paulreece commented:
The main divergence is how it treats the ' ' character. As I noted above, the implementation you suggested here will still add extra spaces in mixed negative/positive arrays. I do like how much cleaner the branching is in your suggestion, I implemented that with a check to convert the sign for all arrays that aren't entirely positive from ' ' to '-'. This implements spacing similar to what the FloatingFormat already does.

#longer printed array with '+' for reference:
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign='+'))
[[+10  -1  +1  +1]
 [+10 +10 +10 +10]]

#version without converting ' ' to '-' on non-positive arrays:
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign='-'))
[[ 10  -1   1   1]
 [ 10  10  10  10]]

#version with conversion ' ' to '-' on non-positive arrays:
>>> print(np.array2string(np.array([[10, -1, 1, 1], [10, 10, 10, 10]]), sign='-'))
[[10 -1  1  1]
 [10 10 10 10]]

#FloatingFormat versions for Reference:
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign='+'))
[[+10.  -1.  +1.  +1.]
 [+10. +10. +10. +10.]]
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign=' '))
[[10. -1.  1.  1.]
 [10. 10. 10. 10.]]
>>> print(np.array2string(np.array([[10., -1, 1, 1], [10, 10, 10, 10]]), sign='-'))
[[10. -1.  1.  1.]
 [10. 10. 10. 10.]]

paulreece commented:
I can build different versions of this to test it if you want, but I think this most recent push should cover the bases. Just let me know.  Thanks again for your feedback.

paulreece commented:
@seberg I just squashed all of my commits into one.  Please let me know if this requires further changes or if I can be of any further assistance.

seberg commented:
Thanks, LGTM to me now.

> I implemented that with a check to convert the sign for all arrays that aren't entirely positive from ' ' to '-'. This implements spacing similar to what the FloatingFormat already does.

True, but that looks wrong.  I don't think we should copy incorrect formatting from float just to be consistent.
